[MIG] l10n_be: tax tags from tax codes

sbidoul opened the pull request:
The current migration of taxes on move lines does not work for l10n_be (and other localizations - see for instance #1172)

This one works for l10n_be and is quite generic, so should work for other localizations.

In a nutshell, it does the following:
* it uses a CSV file that maps tax codes to tax tags, tells if a code is for base or tax, and tells if a positive tax_amount is debit or credit
* from that it creates an inactive tax named "Mig Code {code}"
* it uses these taxes in the migrated move lines

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

sbidoul committed:
[MIG] l10n_be: tax tags from tax codes

pedrobaeza commented:
For activating tests, you have to put on the documentation the module as `Done`. Is it worth to put the general methods elsewhere (on account for example)?

sbidoul commented:
If you think this method (ie mapping old account tax code to new account tags) is useful for other localizations, I can move the method to account, yes.

pedrobaeza commented:
Well, I don't know if it can be useful for others, but it's probably, don't you think?

rousseldenis reviewed (approved):
Tested.

sbidoul commented:
@pedrobaeza where is the best place for such a reusable method? `account/migrations` is not importable.

StefanRijnhart commented:
@sbidoul we don't have it at this point, but conventionally you can create `openerp/openupgrade/openupgrade_90.py` for that.

StefanRijnhart commented:
BTW tests are fixed here: https://github.com/OCA/OpenUpgrade/pull/1434

pedrobaeza commented:
Please rebase and do the documentation stuff

sbidoul committed:
fixup! [MIG] l10n_be: tax tags from tax codes

sbidoul committed:
[WIP] New algorithm to set taxes on move line

sbidoul committed:
fixup! [WIP] New algorithm to set taxes on move line

sbidoul committed:
fixup! [WIP] New algorithm to set taxes on move line

sbidoul committed:
fixup! [WIP] New algorithm to set taxes on move line

sbidoul committed:
fixup! [WIP] New algorithm to set taxes on move line

sbidoul committed:
fixup! [WIP] New algorithm to set taxes on move line

sbidoul commented:
@pedrobaeza @StefanRijnhart I'm now quite satisfied with this algorithm. I encourage you to give it a try.

In short
* you need this table mapping tax codes to tax tags, which should be quite easy to create for each localization
* by default it creates one new inactive tax for each tax code, to ensure the result is both understandable and preserves the vat reports (there is a flag to force it to try and reuse existing taxes, and fall back to creating new taxes if there is an ambiguity)
* it goes to great length to find out if a move line with a tax_code_id is a base or a tax (by looking at the linked invoice)
* it handles debit=credit=0 lines by adding the tax to the corresponding base move line if it can find it reliably (same amount) and falls back to creating two dummy move lines otherwise
* it does not rely on move line labels
* it logs errors for everything that needs to be examined manually 
* it has a method to reset taxes on move lines so you can run it several times in case of problems

sbidoul committed:
fixup! [WIP] New algorithm to set taxes on move line

Optional mechanism to declare if a tax code is base or tax.

sbidoul committed:
fixup! [WIP] New algorithm to set taxes on move line

sbidoul committed:
fixup! [WIP] New algorithm to set taxes on move line

sbidoul committed:
fixup! [WIP] New algorithm to set taxes on move line

sbidoul committed:
fixup! use logged_query

sbidoul committed:
fixup! more logging and no_version=True

sbidoul committed:
fixup! use SQL for inserting 0/0/tax move lines

sbidoul committed:
fixup! v9 is openerp

sbidoul committed:
[FIX] tax migration algorithm, simplify more

It appears that attempting to reuse existing taxes is
a bad idea. Indeed, when running account_chart_update after
migration, tags will be overridden possibly changing past VAT
declarations. So we simply create one inactive tax for each tax tag
we need.

Moreover we provide guidance to determine if a positive value
in tax_amount must be a debit or credit.

pedrobaeza commented:
@sbidoul is this going to be fixed?

pedrobaeza commented:
Closing this for now until further answer.
